src/DataFramesMeta.jl

```diff
@@ -388,19 +387,61 @@ end
 
 function transform(g::GroupedDataFrame; kwargs...)
     result = DataFrame(g)
-    idx2 = cumsum(Int[size(g[i],1) for i in 1:length(g)])
-    idx1 = [1; 1 .+ idx2[1:end-1]]
+    ends = cumsum(Int[size(g[i],1) for i in 1:length(g)])
+    starts = [1; 1 .+ ends[1:end-1]]
     for (k, v) in kwargs
         first = v(g[1])
-        result[k] = Array{eltype(first)}(undef, size(result, 1))
-        result[idx1[1]:idx2[1], k] = first
-        for i in 2:length(g)
-            result[idx1[i]:idx2[i], k] = v(g[i])
+        if first isa AbstractVector 
+            t = Tables.allocatecolumn(eltype(first), size(result, 1))
+            t = _transform!(t, 1, first, first, g, v, starts, ends)
+        else 
+            t = Tables.allocatecolumn(typeof(first), size(result, 1))
+            t = _transform!(t, 1, first, first, g, v, starts, ends)
         end
+        result[k] = t
     end
     return result
 end
 
+function _transform!(t::AbstractVector, i::Int, previous::AbstractVector, current::AbstractVector, g::GroupedDataFrame, v::Function, starts::Vector, ends::Vector)
+    if !(current isa AbstractVector)
+        throw(ArgumentError("Return value must be an `AbstractVector` for all groups or for none of them"))
+    elseif length(current) != size(g[i], 1)
+        throw(ArgumentError("If a function returns a vector, the result " * 
+                            "must have the same length as the groups it operates on"))
+    end
+    S = eltype(current)
+    T = eltype(previous)
+    if !(S <: T || promote_type(S, T) <: T)
+        t = copyto!(Tables.allocatecolumn(promote_type(S, T), length(t)), 
+                    1, t, 1, ends[i-1])
+    end
+    t[starts[i]:ends[i]] = current
+    if i != length(g) 
+        return _transform!(t, i + 1, current, v(g[i+1]), g, v, starts, ends)
```

Comment:
The idea is that you'd call `_transform` only when re-allocating the column. Here you call it for each group, which will trigger a stack overflow for a large number of groups (and is probably slower).

Reply:
Okay. I think I was misunderstanding .

We still have a for-loop. The only difference, really, is that we push the type promotion into a function. Which only ever has to run recursively once or twice.